llvm/tools/sycl-post-link/sycl-post-link.cpp

```diff
@@ -271,6 +271,71 @@ static void collectKernelModuleMap(
   }
 }
 
+// Go through function call graph searching for assert call.
+static bool hasAssertInFunctionCallGraph(llvm::Function *Func) {
+  // Map holds the info about assertions in already examined functions:
+  // true  - if there is an assertion in underlying functions,
+  // false - if there are definetely no assertions in underlying functions.
+  static std::map<llvm::Function *, bool> hasAssertionInCallGraphMap;
+  std::vector<llvm::Function *> FuncCallStack;
+
+  std::vector<llvm::Function *> Workstack;
+  Workstack.push_back(Func);
+
+  while (!Workstack.empty()) {
+    Function *F = Workstack.back();
+    Workstack.pop_back();
+    if (F != Func)
+      FuncCallStack.push_back(F);
+
+    bool IsLeaf = true;
+    for (auto &I : instructions(F)) {
+      if (!isa<CallBase>(&I))
+        continue;
+
+      Function *CF = cast<CallBase>(&I)->getCalledFunction();
+      if (!CF)
+        continue;
+
+      // Return if we've already discovered if there are asserts in the
+      // function call graph.
+      if (hasAssertionInCallGraphMap.count(CF)) {
+        // If we know, that this function does not contain assert, we still
+        // should investigate another instructions in the function.
+        if (!hasAssertionInCallGraphMap[CF])
+          continue;
```

Comment:
This kind of defeats the purpose of the cache, i.e. if we have a huge sub-graph which is known not to have an assertion and it is used from a several kernels, than we would traverse it fully several times.

If I understand correctly, this "if" is needed, because we go upwards by call stack to mark all functions as ones that definitely do not call assert. Technically, "definitely" word is not correct here: if we have a function which calls two other functions and we have only traversed a single call tree and didn't find an assertion, then we can't say that the top-level function _definitely_ don't call assert - it only _might_ not call assert.

A few ideas how to fix this:
- move upwards only until we are sure that all callees of the current function do not call an assertion. As soon as we have encountered a branch in a call three for which the answer is not known, we stop there. Unfortunately, this is likely to require additional data to store or additional actions to perform, so I'm not sure if it worth it.
- during the first traverse, only put `true` values into the map, then perform a second graph traversal and mark everything you haven't marked in the first one with `false`

Since `assert` is an extension, I would expect more kernels/functions without it than kernels/functions with it. Therefore, I think that having an early exit when we know that particular sub-tree does not have an assert is important optimization.

However, this is not a functional problem and if @s-kanaev would like to get this patch sooner, than we can postpone optimizing our graph traversal and do that in a separate PR

Reply:
I'll work on further graph traversal optimization and apply it in a separate PR